Matomo (module, package) updates

Kiwi opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
I got tired of seeing red every time I nixos-rebuilt my server so I updated Matomo. While doing so I found a bug in the latest version that is fixed in the next release (3.12.0) (edit: ...and then the b4 ended up breaking it on NixoOS which stalled this PR for a while... and now ~1 month later and 3.12.0 was released 20 hrs ago :| ) which is currently at beta 3. The project releases beta builds frequently and release builds not so often... so I think it'll be good to be able to use them. Forewarning, you can't downgrade (at least not easily, I don't know how to and I didn't try too hard, I'm sure if you had backups you could...) though as it changes the database. But it's been working better for me than the last one so.

###### Things done
Added a beta build (modeled after the way atom/atom-beta handles it) and fixed deprecation warnings related to phpfpm socket and extraConfig that showed up in nixos-unstable a while ago.

@mmilata added a test for matomo and @aanderse helped me figure out how to make it work on matomo-beta as well. Thanks all!

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @florianjacob

Kiwi committed:
nixos/matomo: fix deprecation warnings

Fixes the phpfpm deprecation warnings about listen and extraConfig by
using fpm.socket and settings. Removes phpfpmProcessManagerConfig.

Kiwi committed:
matomo-beta: init at 3.12.0-b3

Add a beta version of matomo 3.12 that has recent bug fixes. They release these more frequently so it's a good option to have.

aanderse commented on nixos/modules/services/web-apps/matomo.nix:
```diff
@@ -2,15 +2,14 @@
-  # it's not possible to use /run/phpfpm/${pool}.sock because /run/phpfpm/ is root:root 0770,
-  # and therefore is not accessible by the web server.
-  phpSocket = "/run/phpfpm-${pool}.sock";
+  phpSocket = "/run/phpfpm/${pool}.sock";
```
You can get rid of this and use `fpm.socket` instead, if I recall correctly.

Kiwi replied:
When I tried that there or in the place it's used I get this;
`error: attribute 'matomo' missing, at /home/kiwi/nixpkgs/nixos/modules/services/web-apps/matomo.nix:5:9`

aanderse replied:
I don't understand. `phpSocket` is only referenced once throughout `matomo.nix` and it is in documentation. I'll have to checkout your PR and poke around a bit tonight.

aanderse commented on nixos/modules/services/web-apps/matomo.nix:
```diff
@@ -233,15 +213,24 @@ in {
+          "pm.max_spare_servers" = 20;
+          "pm.max_requests" = 500;
+          "catch_workers_output" = true;
+          "env[PIWIK_USER_PATH]" = "${dataDir}" ;
```
-> `phpEnv.PIWIK_USER_PATH = dataDir;`

aanderse commented on nixos/modules/services/web-apps/matomo.nix:
```diff
@@ -233,15 +213,24 @@ in {
-          user = ${user}
-          env[PIWIK_USER_PATH] = ${dataDir}
-          ${cfg.phpfpmProcessManagerConfig}
+        user = "${user}";
```
-> `inherit user;`

aanderse commented on nixos/modules/services/web-apps/matomo.nix:
```diff
@@ -233,15 +213,24 @@ in {
+          log_errors = on
         '';
+        settings = mapAttrs (name: mkDefault) {
+          "listen.owner" = "${socketOwner}";
```
`= "${socketOwner}";` -> `= socketOwner;`

aanderse reviewed (changes requested):
Thanks for taking this on!

aanderse commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15252,7 +15252,9 @@ in
   shaarli-material = callPackage ../servers/web-apps/shaarli/material-theme.nix { };
 
-  matomo = callPackage ../servers/web-apps/matomo { };
+  inherit (callPackage ../servers/web-apps/matomo {})
```
`callPackage` -> `callPackages`

Kiwi replied:
What's the difference? IIRC it seemed kind of arbitrary which one people used for multiple packages.

aanderse commented:
@Kiwi please post configuration you are testing with.

Kiwi commented:
> @Kiwi please post configuration you are testing with.

(Already has it from IRC but for anyone else that may want it) [minimal example](https://gist.github.com/Kiwi/f6629f5aa8c6d25e6f7fed611ee84d6e)

florianjacob commented:
I'll try to have an in-depth look on Sunday. Ahead: Just properly referencing the option instead of the value seems to be a good solution for the `phpSocket` issue.

aanderse commented on nixos/modules/rename.nix:
```diff
@@ -134,6 +134,7 @@ with lib;
     (mkRenamedOptionModule [ "services" "piwik" "enable" ] [ "services" "matomo" "enable" ])
     (mkRenamedOptionModule [ "services" "piwik" "webServerUser" ] [ "services" "matomo" "webServerUser" ])
     (mkRenamedOptionModule [ "services" "piwik" "phpfpmProcessManagerConfig" ] [ "services" "matomo" "phpfpmProcessManagerConfig" ])
```
Out of curiosity what happens if you have `services.piwik.phpfpmProcessManagerConfig` set in your `configuration.nix` with this branch? (ie. an option which has been renamed and then removed)

Kiwi replied:
Yeah I'm not sure if the way I did that is the right way and I don't know this well enough to even know of what other possible approaches there might be. it seemed logical to me, first attempt that fixed the error, seemed to work, and I didn't explore it much after that.

To answer your question...

```
# with services.piwik.*
00:54:19 [kiwi@absurd-nixos nixpkgs]$ sudo nixos-rebuild -I nixpkgs=. build
building Nix...
building the system configuration...
trace: warning: The option definition `services.matomo.phpfpmProcessManagerConfig' in `/home/kiwi/nixpkgs/nixos/modules/rename.nix' no longer has any effect; please remove it.
Use services.phpfpm.pools.<name>.settings
trace: warning: The option `services.piwik.phpfpmProcessManagerConfig' defined in `/etc/nixos/modules/matomo.nix' has been renamed to `services.matomo.phpfpmProcessManagerConfig'.
00:54:34 [kiwi@absurd-nixos nixpkgs]$ sudo vim /etc/nixos/modules/matomo.nix
# changed to services.matomo.*
00:54:58 [kiwi@absurd-nixos nixpkgs]$ sudo nixos-rebuild -I nixpkgs=. build
building Nix...
building the system configuration...
trace: warning: The option definition `services.matomo.phpfpmProcessManagerConfig' in `/etc/nixos/modules/matomo.nix' no longer has any effect; please remove it.
Use services.phpfpm.pools.<name>.settings
00:55:06 [kiwi@absurd-nixos nixpkgs]$
```

aanderse replied:
In that case the message seems confusing so let's modify the `mkRenamedOptionModule` to something like
```
 	    (mkRemovedOptionModule [ "services" "piwik" "phpfpmProcessManagerConfig" ] "Use services.phpfpm.pools.<name>.settings")
```

aanderse commented:
@Kiwi it looks like you have resolved your issues, correct?

Kiwi commented:
> @Kiwi it looks like you have resolved your issues, correct?

Yes, I believe so and then some. (Improved/corrected the documentation more than I had, which also avoids the phpSocket definition altogether)

florianjacob commented:
The code looks good, thanks for getting the code up to date! :) Have yet to test it though.

If you use and feel responsible for the matomo package, feel free to add yourself to the maintainers array to gain some authority when ryan's bot creates an upgrade PR and you tested that it works.

I personally won't commit to maintain beta version in nixpkgs, but I won't object if you do.
As mentioned on IRC, I personally would let the user do something like this (did not test it, but should work), document this in the `package` option, and would think of a helper function to simplify this for the user:

```nix
services.matomo.package = pkgs.matomo.overrideAttrs (oldAttrs: {
    version = "3.12.0b3";
    src = fetchurl {
      url = "https://builds.matomo.org/matomo-${version}.tar.gz";
      sha256 = "1n7b8cag7rpi6y4145cll2irz3in4668jkiicy06wm5nq6lb4bdf";
    };
});
```

aanderse commented:
@florianjacob good point about version... Normally you'd add a package like that to avoid recompiling something heavy (yay binary caches!) but `matomo` is purely script so there isn't any real savings... The argument for is ease of use, but with proper (and visible) documentation it's fairly easy to get an updated version as mentioned. I'll defer to someone else on if the new package commit should be include or not.

florianjacob commented:
@Kiwi when I switch to beta, phpfpm-matomo does not start due to a failure in `matomo-setup-update`. Do you know hat's going on there? Didn't you have a fix for a similar error?

From `journalctl -rb`, i.e. in reverse order:
```
systemd[1]: phpfpm-matomo.service: Job phpfpm-matomo.service/start failed with result 'dependency'.
systemd[1]: Dependency failed for PHP FastCGI Process Manager service for pool matomo.
systemd[1]: Failed to start matomo-setup-update.service.
systemd[1]: matomo-setup-update.service: Failed with result 'exit-code'.
systemd[1]: matomo-setup-update.service: Main process exited, code=exited, status=1/FAILURE
y7mdimnxb23bqdj3j4y2qaffxl1zbj28-unit-script-matomo-setup-update-start[10730]:                                                                                                                                                        
y7mdimnxb23bqdj3j4y2qaffxl1zbj28-unit-script-matomo-setup-update-start[10730]:   The directory "/nix/store/1xszg1fcqsabzj1sk2clc8my7acbmnis-matomo-beta-3.12.0-b4/share/tmp/cache/tracker/" does not exist and could not be created.
y7mdimnxb23bqdj3j4y2qaffxl1zbj28-unit-script-matomo-setup-update-start[10730]:   [InvalidArgumentException]
y7mdimnxb23bqdj3j4y2qaffxl1zbj28-unit-script-matomo-setup-update-start[10730]:                                                                                                                                                        
y7mdimnxb23bqdj3j4y2qaffxl1zbj28-unit-script-matomo-setup-update-start[10730]: The directory "/nix/store/1xszg1fcqsabzj1sk2clc8my7acbmnis-matomo-beta-3.12.0-b4/share/tmp/cache/tracker/" does not exist and could not be created.
systemd[1]: Starting matomo-setup-update.service...
```

mmilata committed:
nixos/matomo: add test

mmilata commented:
@Kiwi Thanks for fixing this! I suck at using github's search and opened duplicate PR ↑. It adds simple NixOS test. Would you be interested in adding it to your PR? Feel free to use it: https://github.com/mmilata/nixpkgs/commit/068c0b27f0d0641173a9b5283ad7c4ff16ab0173. It's missing a maintainer - I'm OK with being one if you're not interested.

mmilata commented on nixos/modules/rename.nix:
```diff
@@ -135,7 +135,7 @@ with lib;
     (mkRenamedOptionModule [ "services" "piwik" "enable" ] [ "services" "matomo" "enable" ])
     (mkRenamedOptionModule [ "services" "piwik" "webServerUser" ] [ "services" "matomo" "webServerUser" ])
-    (mkRenamedOptionModule [ "services" "piwik" "phpfpmProcessManagerConfig" ] [ "services" "matomo" "phpfpmProcessManagerConfig" ])
+    (mkRemovedOptionModule [ "services" "piwik" "phpfpmProcessManagerConfig" ] "Use services.phpfpm.pools.<name>.settings")
```
Shouldn't there be two lines for both `[ "services" "piwik" "phpfpmProcessManagerConfig" ]` and `[ "services" "matomo" "phpfpmProcessManagerConfig" ]` so that people using the latter get more informative error message?

aanderse replied:
Yeah that was the idea... hmm...

aanderse commented:
ping @Kiwi

Kiwi commented:
Apologies this took so long. I hope I've done everything right...

If I didn't mess anything up:

I have reverted the beta back to the last version that worked  which was 3.12.0-b3. I have tried the latest versions... they're still broken. It's up to version 3.12.0-rc2 at this point. I'm not sure how to fix it, either. I've looked briefly but I and I think everyone else that's looked at it have been incredibly busy >.> and I'm not on good (or recent) terms with PHP...

I think I have the removed modules lines how they're supposed to be??

I git cherry-picked the test above that @mmilata supplied.
It could probably be added to the beta build too...

I rebased on to master as of a few minutes ago because there were conflicts with the old PR. I've been watching the bot go, as I type, and it looks like it's probably going to pass (2 to go, 10 passed) Aha! there it is. 10 passed.

I wish I could get it building the newest version which I expect will be out soon...but as it is now, the beta 3 I believe is still better than what we have currently, 3.11.0. I know it's at least fixed one annoying bug that was bothering me in 3.11.0, I don't know either way if b3 introduced any bugs, but so far it's been fine... and I can't go back to 3.11.0 (the beta updated the database) on my server so I'd really appreciate if this was merged. :)

Anyway, if this PR is satisfactory to all involved for what it does?? I think it should be merged. After all it does fix multiple problems; known bugs in 3.11.0 if you choose to use the beta, which is now possible, as well as the phpfpm deprecation warnings being resolved. And it has a test...! :) After that we can worry about getting the next version to work in a separate PR that won't conflict with this one.... as soon as anyone that can fix it has time. and add that other test...

florianjacob commented:
@GrahamcOfBorg test matomo

florianjacob reviewed (approved):
The port to the new phpfpm options is done well and the reviewer feedback was adressed. We should merge this now and do a backport to 19.09, at least for the new phpfpm options. Finding the cause for the issues with the upcoming matomo version will need more time for research.

@Kiwi volunteered to maintain beta versions, so I'm fine with having multiple versions in nixpkgs. This could also be a handy way to package the new stable versions which do not contain security fixes, i.e. for which we don't want to change the default in stable, but to give users the choice to upgrade if they want without relying on unstable.

@mmilata thanks a lot for the test, did not think of such a simple version as I was busy thinking about the impossibility of matomo automated installs, but even just checking the installer page adds a lot of value to see wether the stack generally works.

aanderse commented:
Since @florianjacob has approved there isn't too much to say past this point. If this PR suits him it should be good to go.

The only remaining thing for this PR is we might consider testing the `beta` version of `matomo` as well so when someone files a PR to update it we can merge more confidently. Sound good @Kiwi?

Kiwi commented:
(already addressed on IRC but for anyone else watching) @aanderse Yes, that sounds good. I should have it taken care of either tonight or tomorrow.

aanderse commented:
@GrahamcOfBorg test matomo.matomo matomo.matomo-beta

aanderse commented:
Thanks everyone for your contributions on this.

In a future PR it would be nice to see this module take advantage of the `services.mysql.ensureUsers` option to provide a seamless experience for the user. I would also love to see usage of `PermissionsStartOnly` removed from this module as that feature has been deprecated.

Kiwi commented:
> I would also love to see usage of `PermissionsStartOnly` removed from this module as that feature has been deprecated.

HAHAHAHHA

What the... I'm floored!

This couldn't be more perfect

Let me make sure I understand this correctly...

You're saying that this PR, the primary purpose of which was to fix deprecation warnings, took me so long to wrap up that something else was deprecated in the meantime? Ha! (and if that's not true, please let me remain ignorant, just this once)

Lest us not forget the number of betas (all broken after b3) of 3.12.0 there were. And indeed, less than 24 hours ago, the official version 3.12.0 was released. Which we still can't use.

and of course who can forget about the 2 subsequent duplicate PRs during this time.

This is the best laugh I've had all week. This has gone far beyond being mostly absurd to entirely. I love it. It's amazing. <3

aanderse commented:
`PermissionsStartOnly` was deprecated 12 months ago.

Thanks for your patience on this one. For future reference had this been 3 separate PRs they probably would have been much faster to merge.

Great job everyone! :tada:

Kiwi commented:
> `PermissionsStartOnly` was deprecated 12 months ago.

Boo. :( I reject your reality! Maybe you mean 12 days? Yes, surely it was deprecated 12 days ago, not months.

> Thanks for your patience on this one. For future reference had this been 3 separate PRs they probably would have been much faster to merge.

Yeah, most likely, and in the future I'll keep that in mind... this time it ended up being incredibly amusing and I wouldn't have had to look at how to do the tests if it were 3.

Thanks for all of your help :D

florianjacob commented:
Thanks alot to everyone! I will try to investigate the issues with the new 3.12 release soon.

For the record: I intentionally did not use `ensureUsers` in the module due to the lack of automated install for matomo, I wanted the admin to create database and user by themselves so that they know what to enter. Automatic install has come in reach though, if we manage to pack this plugin, the Matomo installer is a thing of the past for NixOS users: https://plugins.matomo.org/ExtraTools . This will be the next big thing for matomo on NixOS when the current 3.12 issues are fixed.

mmilata commented:
Backport to 19.09: #73692. I've dropped the commit that adds `matomo-beta`, let me know if you want me to add it.
